QualifiedTeachersApi/src/QualifiedTeachersApi/V3/Handlers/GetNpqCertificateHandler.cs

```diff
@@ -0,0 +1,78 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using MediatR;
+using QualifiedTeachersApi.DataStore.Crm;
+using QualifiedTeachersApi.DataStore.Crm.Models;
+using QualifiedTeachersApi.Services.Certificates;
+using QualifiedTeachersApi.V3.Requests;
+using QualifiedTeachersApi.V3.Responses;
+
+namespace QualifiedTeachersApi.V3.Handlers;
+
+public class GetNpqCertificateHandler : IRequestHandler<GetNpqCertificateRequest, GetCertificateResponse>
+{
+    private const string FullNameFormField = "Full Name";
+    private const string PassDateFormField = "Pass Date";
+    private readonly IDataverseAdapter _dataverseAdapter;
+    private readonly ICertificateGenerator _certificateGenerator;
+
+    public GetNpqCertificateHandler(
+        IDataverseAdapter dataverseAdapter,
+        ICertificateGenerator certificateGenerator)
+    {
+        _dataverseAdapter = dataverseAdapter;
+        _certificateGenerator = certificateGenerator;
+    }
+
+    public async Task<GetCertificateResponse> Handle(GetNpqCertificateRequest request, CancellationToken cancellationToken)
+    {
+        var qualification = await _dataverseAdapter.GetQualificationById(
+            request.QualificationId,
+            columnNames: new[]
+            {
+                dfeta_qualification.Fields.dfeta_CompletionorAwardDate,
+                dfeta_qualification.Fields.dfeta_Type,
+                dfeta_qualification.Fields.StateCode,
+                dfeta_qualification.Fields.dfeta_PersonId
+            },
+            contactColumnNames: new[]
+            {
+                Contact.PrimaryIdAttribute,
+                Contact.Fields.FirstName,
+                Contact.Fields.MiddleName,
+                Contact.Fields.LastName
+            });
+
+        if (qualification == null
+            || !qualification.dfeta_CompletionorAwardDate.HasValue
+            || !qualification.dfeta_Type.Value.IsNpq()
+            || qualification.StateCode != dfeta_qualificationState.Active)
+        {
+            return null;
+        }
+
+        var teacher = qualification.Extract<Contact>(Contact.EntityLogicalName, Contact.PrimaryIdAttribute);
+        if (teacher == null)
+        {
+            return null;
+        }
+
+        var fieldValues = new Dictionary<string, string>()
+        {
```

Comment:
Nit: double space if middle name is empty?

Reply:
Interestingly it is the same as currently in the old system!
I'll fix it though as it is not hard

Reply:
done